File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
from collections import deque

def solution(numbers, target):
    q = deque()
    q.append((-1, 0))
    answer =0
    while q:
        index, result = q.popleft()
        
        if index == len(numbers) - 1:
            if result == target:
                answer += 1
            continue
            
        next_index = index + 1
        next_num = numbers[next_index]
        q.append((next_index, result+next_num))
        q.append((next_index, result-next_num))
    return answer       
            
                
        
    
    
    
    

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    public int solution(int[] numbers, int target) {
        return calc(numbers, 0, target, 0);
    }
    
    public int calc(int[] numbers, int index, int target, int current) {
        if (index == numbers.length) {
            return current == target ? 1 : 0;
        }
        
        return calc(numbers, index + 1, target, current + numbers[index]) + calc(numbers, index + 1, target, current - numbers[index]);
    }
}